README.md
# NAME

Perl::Build - perl builder

# SYNOPSIS

# Install as plenv plugin (Recommended)

    % git clone https://github.com/tokuhirom/Perl-Build.git $(plenv root)/plugins/perl-build/

# CLI interface without dependencies

    # perl-build command is FatPacker ready
    % curl -L https://raw.githubusercontent.com/tokuhirom/Perl-Build/master/perl-build | perl - 5.16.2 /opt/perl-5.16/

# CLI interface

    % cpanm Perl::Build
    % perl-build 5.16.2 /opt/perl-5.16/

## Programmable interface

    # install perl from CPAN
    my $result = Perl::Build->install_from_cpan(
        '5.16.2' => (
            dst_path          => '/path/to/perl-5.16.2/',
            configure_options => ['-des'],
        )
    );

    # install perl from tar ball
    my $result = Perl::Build->install_from_tarball(
        'path/to/perl-5.16.2.tar.gz' => (
            dst_path          => '/path/to/perl-5.16.2/',
            configure_options => ['-des'],
        )
    );

# DESCRIPTION

This is yet another perl builder module.

**THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE**.

# METHODS

- `Perl::Build->install_from_cpan($version, %args)`

    Install `$version` perl from CPAN. This method fetches tar ball from CPAN, build, and install it.

    You can pass following options in `%args`.

    - `dst_path`

        Destination directory to install perl.

    - `configure_options : ArrayRef(Optional)`

        Command line arguments for `./Configure`.

        (Default: `['-de']`)

    - `tarball_dir` (Optional)

        Temporary directory to put tar ball.

    - `build_dir` (Optional)

        Temporary directory to build binary.

    - `jobs: Int` (Optional)

        Parallel building and testing.

        (Default: `1`)

- `Perl::Build->install_from_tarball($dist_tarball_path, %args)`

    Install perl from tar ball. This method extracts tar ball, build, and install.

    You can pass following options in `%args`.

    - `dst_path` (Required)

        Destination directory to install perl.

    - `configure_options : ArrayRef` (Optional)

        Command line arguments for `./Configure`.

        (Default: `['-de']`)

    - `build_dir` (Optional)

        Temporary directory to build binary.

    - `jobs: Int` (Optional)

        Parallel building and testing.

        (Default: `1`)

- `Perl::Build->install(%args)`

    Build and install Perl5 from extracted source directory.

    - `src_path` (Required)

        Source code directory to build.  That contains extracted Perl5 source code.

    - `dst_path` (Required)

        Destination directory to install perl.

    - `configure_options : ArrayRef` (Optional)

        Command line arguments for `./Configure`.

        (Default: `['-de']`)

    - `test: Bool` (Optional)

        If you set this value as `true`, `Perl::Build` runs `make test` after building.

        (Default: `0`)

    - `jobs: Int` (Optional)

        Parallel building and testing.

        (Default: `1`)

    Returns an instance of [Perl::Build::Built](https://metacpan.org/pod/Perl%3A%3ABuild%3A%3ABuilt) to facilitate using the built perl from code.

- `Perl::Build->symlink_devel_executables($bin_dir:Str)`

    Perl5 binary generated with `-Dusedevel`, is "perl-5.12.2" form. This method symlinks "perl-5.12.2" to "perl".

# FAQ

- How can I use patchperl plugins?

    If you want to use patchperl plugins, please Google "PERL5\_PATCHPERL\_PLUGIN".

- What's the difference between `perlbrew`?

    [perlbrew](https://metacpan.org/pod/perlbrew) is a perl5 installation manager. But perl-build is a simple perl5 compilation and installation assistant tool.
    It makes perl5 installation easily. That's all. perl-build doesn't care about the user's environment.

    So, perl-build is just an installer.

# THANKS TO

Most of the code was taken from [`App::perlbrew`](https://metacpan.org/pod/App%3A%3Aperlbrew).

TYPESTER - suggests `--patches` option

Thanks

# AUTHOR

Tokuhiro Matsuno <tokuhirom@gmail.com>

# LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This software takes lot of the code from [App::perlbrew](https://metacpan.org/pod/App%3A%3Aperlbrew). App::perlbrew's license is:

    The MIT License

    Copyright (c) 2010,2011 Kang-min Liu

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE.


Changes
Revision history for Perl extension Perl-Build

{{$NEXT}}

1.34 2023-05-16T12:48:45Z
    - Fix document (bes-internal #104)

1.33 2022-05-05T01:28:16Z
    - Use https for github urls (AnaTofuZ #103)

1.32 2021-03-10T15:23:02Z
    - Show perl & Devel::PatchPerl versions in perl-build --version (#98)

1.31 2020-07-12T02:05:29Z
    - Bump HTTP::Tinyish https://github.com/miyagawa/HTTP-Tinyish/pull/18

1.30 2019-10-25T23:43:38Z
    - Use github for blead archive (haarg toddr #94, #95)
    - Fix grammar (scop #92)
    - Fix `plenv install --version` (#91)

1.29 2018-12-21T12:34:40Z
    - Use https explicitly for installing blead perl (Grinnz #89)

1.28 2018-11-20T23:10:18Z
    - Allow installing from local tar.xz (Grinnz #88)

1.27 2018-09-24T03:07:15Z
    - Bump HTTP::Tinyish prereq

1.26 2018-09-23T12:24:24Z
    - Change hardcoded plenv path in README and module pod to $(plenv root) (mikkoi #86)

1.25 2018-08-23T12:18:36Z
    - Dump response body if http status = 599 (#85)

1.24 2018-07-23T14:30:29Z
    - Support perl releases in BackPAN (#84)

1.23 2018-05-27T22:14:34Z
    - Support tar.xz (#81)

1.22 2018-05-27T15:26:27Z
    - Silence "Use of uninitialized value" warning on old perls (#80)

1.21 2018-05-27T14:14:59Z
    - Silence "[cpan_perl_releases] not found the tarball" warning (#76)
    - Remove unnecessary 'use lib "lib"' (#77)
    - Relax perl requirement 5.8.2 -> 5.8.1 (#79)

1.20 2018-05-26T12:28:58Z
    - Resolve perl release versions and URLs from MetaCPAN entirely (Grinnz, sjn #66 #67, #73, #74, #75)
    - Use HTTP::Tinyish for transparent https support (AnaTofuZ #72)
    - If error status is 599 also display content (djzort #71)
    - Add -L to curl in installation instructions (akarelas #68, #70)
    - Support PERL_BUILD_INSTALL_OPTIONS env var (defc0n #61)
    - blead.tar.gz is now extracted into perl-blead-(hash), not perl-(hash) (charsbar #60)

1.13 2016-04-02T05:40:07Z

    commit d757b7f982b8efa48d1b802b7fd91c4ff0c03040
    Merge: 37a22d8 db28ded
    Author: Tatsuhiko Miyagawa <miyagawa@gmail.com>
    Date:   Sat Jan 23 09:05:09 2016 -0800

        Merge pull request #54 from tokuhirom/rc-version

        Remove RC version

    commit db28ded756fe80b092a233d98cbce85a5c9ddd77
    Author: Syohei YOSHIDA <syohex@gmail.com>
    Date:   Sat Jan 23 17:23:16 2016 +0900

        Remove RC version

        Because old RC versions were removed.

    commit 37a22d802074be898c8dcd0c1fa49ccbf20b57a0
    Author: Tokuhiro Matsuno <tokuhirom@gmail.com>
    Date:   Fri Jan 8 19:09:20 2016 +0900

        Rerun fatpacker - Close #52

    commit 445bb6a7cc87ab4c28ca976f0449821b9403082a
    Merge: 320b022 4c2c637
    Author: Tokuhiro Matsuno <tokuhirom@gmail.com>
    Date:   Fri Sep 18 14:47:19 2015 +0900

        Merge pull request #51 from tokuhirom/fix-xz-fetch-issue

        Fix xz fetch issue

    commit 4c2c637578ca8c1d146c873f9494e810f64b39ab
    Author: Syohei YOSHIDA <syohex@gmail.com>
    Date:   Fri Sep 18 14:46:07 2015 +0900

        Update perl-build

    commit 8fc95b0fb9f964f6b202404bc79336080dd63e92
    Author: Syohei YOSHIDA <syohex@gmail.com>
    Date:   Fri Sep 18 14:44:04 2015 +0900

        Fix fetching .xz file issue

1.12 2015-09-15T02:58:10Z

    - support cperl.
    - support installing from https

1.11 2015-07-07T23:25:44Z

    - Support stableperl(moznion)

1.10 2014-09-29T06:54:29Z

    - support PERL_BUILD_COMPILE_OPTIONS env (charsbar)

1.09 2014-09-24T02:46:59Z

    - add support for install blead from perl5.git snapshot
      (charsbar, syohex)

1.08 2014-06-09T02:57:17Z

    - requires newer Pod::Usage for fixing a dependencies error on perl-5.8
      (kazeburo)
    - change shebang of perl-build command
      (kazeburo)

1.07 2014-06-05T07:02:16Z

    - Change default cpan mirror to www.cpan.org.
      (kazeburo++)
    - use perl-releases page instead of scraping search.cpan.org
      (kazeburo++)
    - update github raw content URI
      ref: https://developer.github.com/changes/2014-04-25-user-content-security/
      (chiba)

1.06 2014-04-04T01:11:10Z

    commit e04fcb3280188992fb3ebf85bf93ca11391b08ef
    Author: tokuhirom <tokuhirom@gmail.com>
    Date:   Tue Nov 5 09:02:09 2013 +0900

        Support -D,-A,-U,-j as a first argument.
        https://github.com/tokuhirom/plenv/issues/61

    commit 2c59fa4836d6dab57b1fe079f4e0ca1ff5350f01
    Author: Katsuhiro KONISHI <konishi@hauN.org>
    Date:   Thu Oct 17 17:43:07 2013 +0900

        change /usr/bin/perl to $^X

1.05 2013-09-04T01:46:13Z

    - document `--test` option.
      (tokuhirom)
    - Tweaks needed for conformance with CPAN::Changes::Spec
      (Neil Bowers)
    - plenv-install: Use 'permute' option for Getopt::Long::Parser
      It allows `plenv install 5.18.1 -Dusedtrace --as 5.18.1-dtrace`.
      This issue was reported by hanekomu.
      (tokuhirom)

1.04 2013-08-13T00:03:11Z

    - Added --as option for plenv-install command
      (tokuhirom)
    - fix case for PLENV_ROOT environment variable in plenv-uninstall
      (Akihiro Uchida)
    - Provide plenv completions. Closed #13
      (tokuhirom)

1.03 2013-05-31T00:35:58Z

    - It works well as plugin for plenv.
      (tokuhirom)

1.01 2013-05-29T21:27:28Z

    - Added --symlink-devel-executables option.
      (tokuhirom)
    - Added --version option
      (tokuhirom)

1.00 2013-05-29T09:01:22Z

    [INCOMPATIBLE CHANGES]
    - `--definitions` and #available_versions returns version number without 'perl-' prefix.
      (tokuhirom)

0.21 2013-05-28T14:35:30Z

    - Added `--definitions` option. This option is same as ruby-build.
      (tokuhirom)

0.20 2013-05-17T00:46:48Z

    - use 'test_harness' for parallel testing.
      (tokuhirom)
    - Added --tarball-dir option
      (tokuhirom)

0.19 2013-05-16T02:17:08Z

    - no feature changes

0.18 2013-05-16T01:51:30Z

    - Support --jobs for parallel building.
      (tokuhirom)
    - Install from url
      (tokuhirom)

0.17 2013-04-05T08:36:20Z


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl_5',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => '0.4005',
    },

    requires => {
        'CPAN::Perl::Releases' => '3.58',
        'CPAN::Perl::Releases::MetaCPAN' => '0.006',
        'Devel::PatchPerl' => '0.88',
        'File::Temp' => '0',
        'File::pushd' => '0',
        'Getopt::Long' => '0',
        'HTTP::Tinyish' => '0.17',
        'JSON::PP' => '0',
        'Pod::Usage' => '1.63',
        'perl' => '5.008001',
    },

    recommends => {
    },

    suggests => {
    },

    build_requires => {
    },

    test_requires => {
        'Test::More' => '0.98',
    },

    name            => 'Perl-Build',
    module_name     => 'Perl::Build',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();



author/fatpack.pl
#!/usr/bin/env perl
use strict;
use warnings;
use FindBin;
use App::FatPacker::Simple;
use App::cpm::CLI;
use Carton::Snapshot;
use CPAN::Meta::Requirements;
use Getopt::Long ();

sub cpm {
    App::cpm::CLI->new->run(@_) == 0 or die
}

sub fatpack {
    App::FatPacker::Simple->new->parse_options(@_)->run
}

sub gen_snapshot {
    my $snapshot = Carton::Snapshot->new(path => "cpanfile.snapshot");
    my $no_exclude = CPAN::Meta::Requirements->new;
    $snapshot->find_installs("local", $no_exclude);
    $snapshot->save;
}

Getopt::Long::GetOptions
    "u|update" => \my $update,
    "h|help" => sub { exec "perldoc", $0 },
or exit 1;

chdir $FindBin::Bin;

my $target = '5.8.1';

my $resolver = -f "cpanfile.snapshot" && !$update ? "snapshot" : "metadb";

warn "Resolver $resolver\n";
cpm "install", "--cpanfile", "../cpanfile", "--target-perl", $target, "--resolver", $resolver;
cpm "install", "--resolver", $resolver, "--reinstall", "ExtUtils::MakeMaker";
gen_snapshot if $update;
print STDERR "FatPacking...";
fatpack
    "-q",
    "-o", "../perl-build",
    "-d", "local,../lib",
    "-e", "Test::Simple,Test,File::Spec,Carp",
    "--shebang", '#!/usr/bin/env perl',
    "../script/perl-build";
print STDERR " DONE\n";

__END__

=head1 NAME

fatpack.pl - fatpack perl-build

=head1 SYNOPSIS

  perl fatpack.pl
  perl fatpack.pl --update

=head1 DESCRIPTION

This script does:

=over 4

=item *

install CPAN module dependencies of perl-build into ./local with cpanfile.snapshot

=item *

fatpack ../script/perl-build with modules in ./local and ../lib

=back

If you want to update CPAN module dependencies and re-generate cpanfile.snapshot,
then execute this script with C<--update> option.

=head1 REQUIREMENT

App::cpm, App::FatPacker::Simple, Carton

You can install them by

  cpanm -nq App::cpm App::FatPacker::Simple Carton

=cut



lib/Perl/Build.pm
package Perl::Build;
use strict;
use warnings;
use utf8;

use 5.008001;
our $VERSION = '1.34';

use Carp ();
use File::Basename;
use File::Spec::Functions qw(catfile catdir rel2abs);
use CPAN::Perl::Releases;
use CPAN::Perl::Releases::MetaCPAN;
use File::pushd qw(pushd);
use File::Temp;
use HTTP::Tinyish;
use JSON::PP qw(decode_json);
use Devel::PatchPerl 0.88;
use Perl::Build::Built;
use Time::Local;

our $CPAN_MIRROR = $ENV{PERL_BUILD_CPAN_MIRROR} || 'https://cpan.metacpan.org';

sub available_perls {
    my $class = shift;

    my $releases = CPAN::Perl::Releases::MetaCPAN->new->get;
    my @available_versions;
    for my $release (@$releases) {
        if ($release->{name} =~ /^perl-(5.(\d+).(\d+)(-\w+)?)$/) {
            my ($version, $major, $minor, $rc) = ($1, $2, $3, $4);
            my $sort_by = sprintf "%03d%03d%s", $major, $minor, $rc || "ZZZ";
            push @available_versions, { version => $version, sort_by => $sort_by };
        }
    }
    map { $_->{version} } sort { $b->{sort_by} cmp $a->{sort_by} } @available_versions;
}

# @return extracted source directory
sub extract_tarball {
    my ($class, $dist_tarball, $destdir) = @_;

    # Was broken on Solaris, where GNU tar is probably
    # installed as 'gtar' - RT #61042
    my $tar = $^O eq 'solaris' ? 'gtar' : 'tar';

    my $type
        = $dist_tarball =~ m/bz2$/  ? 'j'
        : $dist_tarball =~ m/xz$/   ? 'J'
                                    : 'z';

    my $abs_tarball = File::Spec->rel2abs($dist_tarball);

    my @tar_files = `$tar t${type}f "$abs_tarball"`;
    $? == 0
        or die "Failed to extract $dist_tarball";

    chomp @tar_files;
    my %seen;
    my @prefixes = grep !$seen{$_}++, map m{\A(?:\./)?([^/]+)}, @tar_files;

    die "$dist_tarball does not contain single directory : @prefixes"
        if @prefixes != 1;

    system(qq{cd "$destdir"; $tar x${type}f "$abs_tarball"}) == 0
        or die "Failed to extract $dist_tarball";

    return catfile($destdir, $prefixes[0]);
}

sub perl_release {
    my ($class, $version) = @_;

    my ($dist_tarball, $dist_tarball_url);
    my @err;
    for my $func (qw/cpan_perl_releases metacpan/) {
        eval {
            ($dist_tarball, $dist_tarball_url) = $class->can("perl_release_by_$func")->($class,$version);
        };
        push @err, "[$func] $@" if $@;
        last if $dist_tarball && $dist_tarball_url;
    }
    if (!$dist_tarball and !$dist_tarball_url) {
        push @err, "ERROR: Cannot find the tarball for perl-$version\n";
        die join "", @err;
    }

    return ($dist_tarball, $dist_tarball_url);
}

sub perl_release_by_cpan_perl_releases {
    my ($class, $version) = @_;
    my $tarballs = CPAN::Perl::Releases::perl_tarballs($version);
    my $x = $tarballs->{'tar.gz'} || $tarballs->{'tar.bz2'} || $tarballs->{'tar.xz'};
    die "not found the tarball for perl-$version\n" unless $x;
    my $dist_tarball = (split("/", $x))[-1];
    my $dist_tarball_url = $CPAN_MIRROR . "/authors/id/$x";
    return ($dist_tarball, $dist_tarball_url);
}

sub perl_release_by_metacpan {
    my ($class, $version) = @_;
    my $releases = CPAN::Perl::Releases::MetaCPAN->new->get;
    for my $release (@$releases) {
        if ($release->{name} eq "perl-$version") {
            my ($path) = $release->{download_url} =~ m{(/authors/id/.*)};
            my $dist_tarball = (split("/", $path))[-1];
            my $dist_tarball_url = $CPAN_MIRROR . $path;
            return ($dist_tarball, $dist_tarball_url);
        }
    }
    die "not found the tarball for perl-$version\n";
}

sub perl_release_by_github {
    my ($class, $committish) = @_;

    # URL encode the committish
    require HTTP::Tiny;
    ( my $c = HTTP::Tiny->www_form_urlencode([''=>$committish]) ) =~ s/^=//;
    my $commit_url = "https://api.github.com/repos/Perl/perl5/commits/$c";

    # The API puts the error message in the content, so we can't http_get
    my $res = HTTP::Tinyish->new(verify_SSL => 1)->get($commit_url);

    my %content = do { local $@; local $SIG{__DIE__}; eval {
        %{ decode_json( $res->{content} ) } } };

    unless ($res->{success}) {
        my $m = $content{message} || $res->{reason};
        die "$m\n";
    }

    my $name = $committish;
    $committish = $content{sha};

    # If they specified part of the commit-hash use the full hash as the name.
    # Otherwise they specified a branch or tag, like "blead" or "blead@{date}"
    # and we'll use the prefix plus the commit date.

    if ( $committish =~ /^\Q$name/i ) {
        $name = $content{sha};
    } else {
        $name =~ s/@.*$//;
        $name =~ s{/}{_}g; # branches can have slashes, dest can't
        ( my $date = $content{commit}{committer}{date} ) =~ s/\D//g;
        $name .= "-$date";
    }

    my $url = "https://github.com/Perl/perl5/archive/$committish.tar.gz";

    return ( $url, $name );
}

sub http_get {
    my ($url) = @_;

    my $http = HTTP::Tinyish->new(verify_SSL => 1);
    my $response = $http->get($url);
    if ($response->{success}) {
        return $response->{content};
    } else {
        my $msg = $response->{status} == 599 ? ", $response->{content}" : "";
        chomp $msg;
        die "Cannot get content from $url: $response->{status} $response->{reason}$msg\n";
    }
}

sub http_mirror {
    my ($url, $path) = @_;

    my $http = HTTP::Tinyish->new(verify_SSL => 1);
    my $response = $http->mirror($url, $path);
    if ($response->{success}) {
        print "Downloaded $url to $path.\n";
    } else {
        my $msg = $response->{status} == 599 ? ", $response->{content}" : "";
        chomp $msg;
        die "Cannot get file from $url: $response->{status} $response->{reason}$msg";
    }
}

sub install_from_cpan {
    my ($class, $version, %args) = @_;

    $args{patchperl} && Carp::croak "The patchperl argument was deprected.";

    my $tarball_dir = $args{tarball_dir}
        || File::Temp::tempdir( CLEANUP => 1 );
    my $build_dir = $args{build_dir}
        || File::Temp::tempdir( CLEANUP => 1 );
    my $dst_path = $args{dst_path}
        or die "Missing mandatory parameter: dst_path";
    my $configure_options = $args{configure_options}
        || ['-de'];

    # download tar ball
    my ( $dist_tarball, $dist_tarball_url )
        = ref $version ? @{$version} : Perl::Build->perl_release($version);
    my $dist_tarball_path = catfile($tarball_dir, $dist_tarball);
    if (-f $dist_tarball_path) {
        print "Use the previously fetched ${dist_tarball}\n";
    }
    else {
        print "Fetching $version as $dist_tarball_path ($dist_tarball_url)\n";
        http_mirror( $dist_tarball_url, $dist_tarball_path );
    }

    # and extract tar ball.
    my $dist_extracted_path = Perl::Build->extract_tarball($dist_tarball_path, $build_dir);
    Perl::Build->install(
        src_path          => $dist_extracted_path,
        dst_path          => $dst_path,
        configure_options => $configure_options,
        test              => $args{test},
        jobs              => $args{jobs},
    );
}

sub install_from_url {
    my ($class, $dist_tarball_url, %args) = @_;
    $args{patchperl} && Carp::croak "The patchperl argument was deprected.";

    my $build_dir = $args{build_dir}
        || File::Temp::tempdir( CLEANUP => 1 );
    my $tarball_dir = $args{tarball_dir}
        || File::Temp::tempdir( CLEANUP => 1 );
    my $dst_path = $args{dst_path}
        or die "Missing mandatory parameter: dst_path";
    my $configure_options = $args{configure_options}
        || ['-de'];

    my $dist_tarball = basename($dist_tarball_url);
    my $dist_tarball_path = catfile($tarball_dir, $dist_tarball);
    if (-f $dist_tarball_path) {
        print "Use the previously fetched ${dist_tarball}\n";
    }
    else {
        print "Fetching $dist_tarball_path ($dist_tarball_url)\n";
        http_mirror( $dist_tarball_url, $dist_tarball_path );
    }

    my $dist_extracted_path = Perl::Build->extract_tarball($dist_tarball_path, $build_dir);
    Perl::Build->install(
        src_path          => $dist_extracted_path,
        dst_path          => $dst_path,
        configure_options => $configure_options,
        test              => $args{test},
        jobs              => $args{jobs},
    );
}

sub install_from_tarball {
    my ($class, $dist_tarball_path, %args) = @_;
    $args{patchperl} && Carp::croak "The patchperl argument was deprected.";

    my $build_dir = $args{build_dir}
        || File::Temp::tempdir( CLEANUP => 1 );
    my $dst_path = $args{dst_path}
        or die "Missing mandatory parameter: dst_path";
    my $configure_options = $args{configure_options}
        || ['-de'];

    my $dist_extracted_path = Perl::Build->extract_tarball($dist_tarball_path, $build_dir);
    Perl::Build->install(
        src_path          => $dist_extracted_path,
        dst_path          => $dst_path,
        configure_options => $configure_options,
        test              => $args{test},
        jobs              => $args{jobs},
    );
}

sub install {
    my ($class, %args) = @_;
    $args{patchperl} && Carp::croak "The patchperl argument was deprected.";

    my $src_path = $args{src_path}
        or die "Missing mandatory parameter: src_path";
    my $dst_path = $args{dst_path}
        or die "Missing mandatory parameter: dst_path";
    my $configure_options = $args{configure_options}
        or die "Missing mandatory parameter: configure_options";
    my $jobs = $args{jobs}; # optional
    my $test = $args{test}; # optional

    unshift @$configure_options, qq(-Dprefix=$dst_path);

    # Perl5 installs public executable scripts(like `prove`) to /usr/local/share/
    # if it exists.
    #
    # This -A'eval:scriptdir=$prefix/bin' option avoid this feature.
    unless (grep { /eval:scriptdir=/} @$configure_options) {
        push @$configure_options, "-A'eval:scriptdir=${dst_path}/bin'";
    }

    # clean up environment
    delete $ENV{$_} for qw(PERL5LIB PERL5OPT);

    {
        my $dir = pushd($src_path);

        # determine_version is a public API.
        my $dist_version = Devel::PatchPerl->determine_version();
        print "Configuring perl '$dist_version'\n";

        # clean up
        $class->do_system("rm -f config.sh Policy.sh");

        # apply patches
        Devel::PatchPerl->patch_source();

        # configure
        $class->do_system(['sh', 'Configure', @$configure_options]);
        # patch for older perls
        # XXX is this needed? patchperl do this?
        # if (Perl::Build->perl_version_to_integer($dist_version) < Perl::Build->perl_version_to_integer( '5.8.9' )) {
        #     $class->do_system("$^X -i -nle 'print unless /command-line/' makefile x2p/makefile");
        # }

        # build
        my @make = qw(make);
        if ($ENV{PERL_BUILD_COMPILE_OPTIONS}) {
            push @make, $ENV{PERL_BUILD_COMPILE_OPTIONS};
        }
        if ($jobs) {
            push @make, '-j', $jobs;
        }
        $class->do_system(\@make);
        if ($test) {
            local $ENV{TEST_JOBS} = $jobs if $jobs;
            # Test via "make test_harness" if available so we'll get
            # automatic parallel testing via $HARNESS_OPTIONS. The
            # "test_harness" target was added in 5.7.3, which was the last
            # development release before 5.8.0.
            my $test_target = 'test';
            if ($dist_version && $dist_version =~ /^5\.([0-9]+)\.([0-9]+)/
                && ($1 >= 8 || $1 == 7 && $2 == 3)) {
                $test_target = "test_harness";
            }
            $class->do_system([@make, $test_target]);
        }
	@make = qw(make install);
	if ($ENV{PERL_BUILD_INSTALL_OPTIONS}) {
	    push @make, $ENV{PERL_BUILD_INSTALL_OPTIONS};
	}
        $class->do_system(\@make);
    }
    return Perl::Build::Built->new({
        installed_path => $dst_path,
    });
}

sub do_system {
    my ($class, $cmd) = @_;

    if (ref $cmd eq 'ARRAY') {
        $class->info(join(' ', @$cmd));
        system(@$cmd) == 0
            or die "Installation failure: @$cmd";
    } else {
        $class->info($cmd);
        system($cmd) == 0
            or die "Installation failure: $cmd";
    }
}

sub symlink_devel_executables {
    my ($class, $bin_dir) = @_;

    for my $executable (glob("$bin_dir/*")) {
        my ($name, $version) = basename( $executable ) =~ m/(.+?)(5\.\d.*)?$/;
        if ($version) {
            my $cmd = "ln -fs $executable $bin_dir/$name";
            $class->info($cmd);
            system($cmd);
        }
    }
}

sub info {
    my ($class, @msg) = @_;
    print @msg, "\n";
}

1;
__END__

=encoding utf8

=for stopwords tarball Optional symlinks patchperl

=head1 NAME

Perl::Build - perl builder

=head1 SYNOPSIS

=head1 Install as plenv plugin (Recommended)

    % git clone https://github.com/tokuhirom/Perl-Build.git $(plenv root)/plugins/perl-build/

=head1 CLI interface without dependencies

    # perl-build command is FatPacker ready
    % curl -L https://raw.githubusercontent.com/tokuhirom/Perl-Build/master/perl-build | perl - 5.16.2 /opt/perl-5.16/

=head1 CLI interface

    % cpanm Perl::Build
    % perl-build 5.16.2 /opt/perl-5.16/

=head2 Programmable interface

    # install perl from CPAN
    my $result = Perl::Build->install_from_cpan(
        '5.16.2' => (
            dst_path          => '/path/to/perl-5.16.2/',
            configure_options => ['-des'],
        )
    );

    # install perl from tar ball
    my $result = Perl::Build->install_from_tarball(
        'path/to/perl-5.16.2.tar.gz' => (
            dst_path          => '/path/to/perl-5.16.2/',
            configure_options => ['-des'],
        )
    );

=head1 DESCRIPTION

This is yet another perl builder module.

B<THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE>.

=head1 METHODS

=over 4

=item C<< Perl::Build->install_from_cpan($version, %args) >>

Install C<< $version >> perl from CPAN. This method fetches tar ball from CPAN, build, and install it.

You can pass following options in C<< %args >>.

=over 4

=item C<< dst_path >>

Destination directory to install perl.

=item C<< configure_options : ArrayRef(Optional) >>

Command line arguments for C<< ./Configure >>.

(Default: C<< ['-de'] >>)

=item C<< tarball_dir >> (Optional)

Temporary directory to put tar ball.

=item C<< build_dir >> (Optional)

Temporary directory to build binary.

=item C<< jobs: Int >> (Optional)

Parallel building and testing.

(Default: C<1>)

=back

=item C<< Perl::Build->install_from_tarball($dist_tarball_path, %args) >>

Install perl from tar ball. This method extracts tar ball, build, and install.

You can pass following options in C<< %args >>.

=over 4

=item C<< dst_path >> (Required)

Destination directory to install perl.

=item C<< configure_options : ArrayRef >> (Optional)

Command line arguments for C<< ./Configure >>.

(Default: C<< ['-de'] >>)

=item C<< build_dir >> (Optional)

Temporary directory to build binary.

=item C<< jobs: Int >> (Optional)

Parallel building and testing.

(Default: C<1>)

=back

=item C<< Perl::Build->install(%args) >>

Build and install Perl5 from extracted source directory.

=over 4

=item C<< src_path >> (Required)

Source code directory to build.  That contains extracted Perl5 source code.

=item C<< dst_path >> (Required)

Destination directory to install perl.

=item C<< configure_options : ArrayRef >> (Optional)

Command line arguments for C<< ./Configure >>.

(Default: C<< ['-de'] >>)

=item C<< test: Bool >> (Optional)

If you set this value as C<< true >>, C<< Perl::Build >> runs C<< make test >> after building.

(Default: C<0>)

=item C<< jobs: Int >> (Optional)

Parallel building and testing.

(Default: C<1>)

=back

Returns an instance of L<Perl::Build::Built> to facilitate using the built perl from code.

=item C<< Perl::Build->symlink_devel_executables($bin_dir:Str) >>

Perl5 binary generated with C<< -Dusedevel >>, is "perl-5.12.2" form. This method symlinks "perl-5.12.2" to "perl".

=back

=head1 FAQ

=over 4

=item How can I use patchperl plugins?

If you want to use patchperl plugins, please Google "PERL5_PATCHPERL_PLUGIN".

=item What's the difference between C<< perlbrew >>?

L<perlbrew> is a perl5 installation manager. But perl-build is a simple perl5 compilation and installation assistant tool.
It makes perl5 installation easily. That's all. perl-build doesn't care about the user's environment.

So, perl-build is just an installer.

=back

=head1 THANKS TO

Most of the code was taken from L<< C<App::perlbrew> >>.

TYPESTER - suggests C<< --patches >> option

Thanks

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom@gmail.comE<gt>


=head1 LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This software takes lot of the code from L<App::perlbrew>. App::perlbrew's license is:

    The MIT License

    Copyright (c) 2010,2011 Kang-min Liu

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE.


